File: web/src/pages/internal/livestreamsTypes.ts
/**
 * Types for the Live streaming control center
 * (GET /api/v1/livestreams?workspace_id=N).
 *
 * Mirrors the backend `livestreamResponse` DTO
 * (pkg/api/livestreams_types.go). The YouTube resource references
 * (broadcast/stream ids) and the stream name/key are deliberately not
 * exposed by the API, so they never appear here either.
 */
export type LivestreamRow = {
  id: string;
  workspace_id: number;
  platform_account_id: number;
  /** Display name of the bound YouTube channel; empty when unknown. */
  channel_name?: string;
  title: string;
  description: string;
  privacy_status: string; // private | unlisted | public
  playback_mode: string; // loop_continuous | play_once
  schedule_type: string; // manual | now | scheduled | recurring
  scheduled_start_at?: string | null;
  desired_state: string;
  actual_state: string;
  resolution: string; // 720p30 | 1080p30
  frame_rate: number;
  auto_restart: boolean;
  /** YouTube numeric category id (empty = none). */
  category: string;
  made_for_kids: boolean;
  /** ISO 639-1 / BCP-47 code (empty = none). */
  language: string;
  /** media_assets.id of the uploaded cover (wizard Carica immagine). */
  thumbnail_media_id?: string | null;
  dvr_enabled: boolean;
  auto_start: boolean;
  auto_stop: boolean;
  /** normal | low | ultraLow */
  latency_preference: string;
  created_at: string;
  updated_at: string;
};

export type LivestreamsResponse = {
  items: LivestreamRow[];
  next_cursor?: string;
  has_more?: boolean;
};

/**
 * Per-channel preflight row from GET /api/v1/livestreams/channels
 * (creation-wizard step 1). LiveEnabled means the persisted grant
 * carries a YouTube live scope (youtube / youtube.force-ssl) — the
 * necessary condition for the Live Streaming API.
 */
export type LivestreamChannel = {
  platform_account_id: number;
  username: string;
  platform_user_id: string;
  account_state: "valid" | "reconnect_required" | "suspended" | "deleted" | string;
  oauth_ready: boolean;
  live_enabled: boolean;
  last_verified_at?: string | null;
  active_lives: number;
};

export type LivestreamChannelsResponse = {
  channels: LivestreamChannel[];
};

export type LivestreamTab =
  | "all"
  | "live"
  | "scheduled"
  | "drafts"
  | "ended"
  | "errors";

export type LivestreamSummary = {
  live: number;
  scheduled: number;
  reconnecting: number;
  errors: number;
};

/**
 * One row of GET /api/v1/media — the Media Library entry the live
 * wizard's step 3 renders. Probe fields stay null until the upload
 * worker probes the asset (ffprobe); `live_compatibility` is derived
 * server-side so the frontend never re-implements the profile check.
 */
export type MediaLibraryItem = {
  id: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  created_at: string;
  /** Short-lived presigned GET URL (15 min) for `<video>` preview. */
  preview_url?: string;
  duration_seconds?: number | null;
  width?: number | null;
  height?: number | null;
  fps?: number | null;
  has_audio?: boolean | null;
  video_codec?: string;
  audio_codec?: string;
  probed_at?: string | null;
  live_compatibility: "ready" | "needs_normalization" | "unknown";
};

export type MediaLibraryResponse = {
  items: MediaLibraryItem[];
  next_cursor?: string;
  has_more?: boolean;
};

/**
 * Step-2 wizard form state (Configurazione YouTube). Lifted to the
 * page so back → forward preserves the entries; steps 3–5 (contenuti,
 * riproduzione, riepilogo) consume it when they land.
 */
export type LivestreamStep2Form = {
  title: string;
  description: string;
  privacy: "private" | "unlisted" | "public";
  category: string; // YouTube numeric id, "" = none
  madeForKids: boolean;
  language: string; // ISO 639-1 code, "" = none
  thumbnailMediaId: string | null;
  dvr: boolean;
  autoStart: boolean;
  autoStop: boolean;
  latency: "normal" | "low" | "ultraLow";
};

export const EMPTY_STEP2_FORM: LivestreamStep2Form = {
  title: "",
  description: "",
  privacy: "unlisted",
  category: "",
  madeForKids: false,
  language: "",
  thumbnailMediaId: null,
  dvr: false,
  autoStart: false,
  autoStop: false,
  latency: "normal",
};

/**
 * Known YouTube video category ids (videoCategories.list, default
 * global region). Static list for the wizard select — YouTube remains
 * the authority at broadcast creation time.
 */
export const YOUTUBE_CATEGORIES: Array<{ id: string; label: string }> = [
  { id: "1", label: "Film e animazione" },
  { id: "2", label: "Auto e veicoli" },
  { id: "10", label: "Musica" },
  { id: "15", label: "Animali domestici" },
  { id: "17", label: "Sport" },
  { id: "18", label: "Cortometraggi" },
  { id: "19", label: "Viaggi e eventi" },
  { id: "20", label: "Gaming" },
  { id: "21", label: "Video blog" },
  { id: "22", label: "Persone e blog" },
  { id: "23", label: "Commedia" },
  { id: "24", label: "Intrattenimento" },
  { id: "25", label: "Notizie e politica" },
  { id: "26", label: "Istruzione e tutorial" },
  { id: "27", label: "Educazione" },
  { id: "28", label: "Scienza e tecnologia" },
  { id: "29", label: "Non profit e attivismo" },
  { id: "42", label: "Shorts" },
];

/**
 * ISO 639-1 languages offered by the wizard (matches the app's
 * language-flag set plus the most common YouTube live languages).
 */
export const LIVESTREAM_LANGUAGES: Array<{ code: string; label: string }> = [
  { code: "it", label: "Italiano" },
  { code: "en", label: "Inglese" },
  { code: "es", label: "Spagnolo" },
  { code: "fr", label: "Francese" },
  { code: "de", label: "Tedesco" },
  { code: "pl", label: "Polacco" },
  { code: "ru", label: "Russo" },
  { code: "id", label: "Indonesiano" },
  { code: "tr", label: "Turco" },
  { code: "hi", label: "Hindi" },
];

/** liveBroadcast.contentDetails.latencyPreference options. */
export const LATENCY_OPTIONS: Array<{
  value: "normal" | "low" | "ultraLow";
  label: string;
  description: string;
}> = [
  { value: "normal", label: "Normale", description: "Latenza standard, qualità massima" },
  { value: "low", label: "Bassa", description: "~10s di ritardo, per interazioni live" },
  { value: "ultraLow", label: "Ultra bassa", description: "~5s di ritardo, per gaming e sport" },
];

File: pkg/api/livestreams_types.go
package api

import (
	"context"
	"time"

	"github.com/Marcuss-ops/InstaeditLogin/internal/models"
)

// LivestreamStore is the persistence contract the livestream CRUD
// handlers depend on. The API keeps this contract local so handler
// tests can use an in-memory fake while production wires the SQL
// repository (internal/repository.LivestreamRepository declares an
// identical interface for its own consumers).
type LivestreamStore interface {
	Create(ctx context.Context, ls *models.Livestream) error
	FindByID(ctx context.Context, id string) (*models.Livestream, error)
	ListByWorkspace(ctx context.Context, workspaceID int64) ([]models.Livestream, error)
	Update(ctx context.Context, ls *models.Livestream) error
	Delete(ctx context.Context, id string) error
}

// createLivestreamRequest is the body accepted by
// POST /api/v1/livestreams. It creates a live CONFIGURATION row in
// state draft — it does not start a broadcast (that is the prepare /
// start control surface).
type createLivestreamRequest struct {
	WorkspaceID       int64   `json:"workspace_id"`
	PlatformAccountID int64   `json:"platform_account_id"`
	Title             string  `json:"title"`
	Description       string  `json:"description,omitempty"`
	PrivacyStatus     string  `json:"privacy_status"`
	PlaybackMode      string  `json:"playback_mode"`
	ScheduleType      string  `json:"schedule_type"`
	ScheduledStartAt  *string `json:"scheduled_start_at,omitempty"` // RFC3339
	Resolution        string  `json:"resolution,omitempty"`
	FrameRate         int     `json:"frame_rate,omitempty"`
	AutoRestart       *bool   `json:"auto_restart,omitempty"`
	// YouTube broadcast metadata (wizard step 2). All optional with
	// documented defaults; the worker maps them onto the liveBroadcast
	// resource at prepare time.
	Category          string  `json:"category,omitempty"`
	MadeForKids       *bool   `json:"made_for_kids,omitempty"`
	Language          string  `json:"language,omitempty"`
	ThumbnailMediaID  *string `json:"thumbnail_media_id,omitempty"`
	DVREnabled        *bool   `json:"dvr_enabled,omitempty"`
	AutoStart         *bool   `json:"auto_start,omitempty"`
	AutoStop          *bool   `json:"auto_stop,omitempty"`
	LatencyPreference string  `json:"latency_preference,omitempty"`
}

// patchLivestreamRequest is the body accepted by
// PATCH /api/v1/livestreams/{id}. Every field is a pointer so absent
// fields stay untouched (read-modify-write partial update; the write
// is last-writer-wins per column).
//
// scheduled_start_at semantics: omit to leave it unchanged, send an
// EMPTY string ("") to clear it, or a valid RFC3339 timestamp to set
// it. JSON null is treated as "not provided" (same as omitting it).
type patchLivestreamRequest struct {
	Title            *string `json:"title,omitempty"`
	Description      *string `json:"description,omitempty"`
	PrivacyStatus    *string `json:"privacy_status,omitempty"`
	PlaybackMode     *string `json:"playback_mode,omitempty"`
	ScheduleType     *string `json:"schedule_type,omitempty"`
	ScheduledStartAt *string `json:"scheduled_start_at,omitempty"` // RFC3339; "" clears
	Resolution       *string `json:"resolution,omitempty"`
	FrameRate        *int    `json:"frame_rate,omitempty"`
	AutoRestart      *bool   `json:"auto_restart,omitempty"`
	// YouTube broadcast metadata (wizard step 2). thumbnail_media_id
	// semantics mirror scheduled_start_at: omit to leave unchanged,
	// send "" to clear, or a media asset id to set.
	Category          *string `json:"category,omitempty"`
	MadeForKids       *bool   `json:"made_for_kids,omitempty"`
	Language          *string `json:"language,omitempty"`
	ThumbnailMediaID  *string `json:"thumbnail_media_id,omitempty"` // "" clears
	DVREnabled        *bool   `json:"dvr_enabled,omitempty"`
	AutoStart         *bool   `json:"auto_start,omitempty"`
	AutoStop          *bool   `json:"auto_stop,omitempty"`
	LatencyPreference *string `json:"latency_preference,omitempty"`
	// desired_state / actual_state are worker-owned; a request that
	// attempts to set them is rejected by the handler.
	DesiredState *string `json:"desired_state,omitempty"`
	ActualState  *string `json:"actual_state,omitempty"`
}

// livestreamResponse is the public shape of a livestream row. The
// YouTube resource references (broadcast/stream ids) and the stream
// name/key are deliberately excluded — the stream name is a technical
// credential that must never reach the frontend. ChannelName is the
// display name of the bound YouTube channel, resolved by the handlers
// from the platform account (empty when the account cannot be loaded).
type livestreamResponse struct {
	ID                string     `json:"id"`
	WorkspaceID       int64      `json:"workspace_id"`
	PlatformAccountID int64      `json:"platform_account_id"`
	ChannelName       string     `json:"channel_name"`
	Title             string     `json:"title"`
	Description       string     `json:"description"`
	PrivacyStatus     string     `json:"privacy_status"`
	PlaybackMode      string     `json:"playback_mode"`
	ScheduleType      string     `json:"schedule_type"`
	ScheduledStartAt  *time.Time `json:"scheduled_start_at,omitempty"`
	DesiredState      string     `json:"desired_state"`
	ActualState       string     `json:"actual_state"`
	Resolution        string     `json:"resolution"`
	FrameRate         int        `json:"frame_rate"`
	AutoRestart       bool       `json:"auto_restart"`
	Category          string     `json:"category"`
	MadeForKids       bool       `json:"made_for_kids"`
	Language          string     `json:"language"`
	ThumbnailMediaID  *string    `json:"thumbnail_media_id,omitempty"`
	DVREnabled        bool       `json:"dvr_enabled"`
	AutoStart         bool       `json:"auto_start"`
	AutoStop          bool       `json:"auto_stop"`
	LatencyPreference string     `json:"latency_preference"`
	CreatedAt         time.Time  `json:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at"`
}

type listLivestreamsResponse struct {
	Items []livestreamResponse `json:"items"`
}

// livestreamChannelResponse is the per-channel preflight row for the
// creation wizard (GET /api/v1/livestreams/channels). OAuthReady and
// LiveEnabled are derived from the persisted grant: LiveEnabled means
// the grant carries a YouTube live scope (youtube or youtube.force-ssl),
// which is the necessary condition for the Live Streaming API — the
// authoritative liveStreamingNotEnabled check still happens on
// broadcast creation. LastVerifiedAt is the account's last_validated_at.
type livestreamChannelResponse struct {
	PlatformAccountID int64        `json:"platform_account_id"`
	Username          string       `json:"username"`
	PlatformUserID    string       `json:"platform_user_id"`
	AccountState      AccountState `json:"account_state"`
	OAuthReady        bool         `json:"oauth_ready"`
	LiveEnabled       bool         `json:"live_enabled"`
	LastVerifiedAt    *time.Time   `json:"last_verified_at,omitempty"`
	ActiveLives       int          `json:"active_lives"`
}

type listLivestreamChannelsResponse struct {
	Channels []livestreamChannelResponse `json:"channels"`
}

func toLivestreamResponse(ls *models.Livestream) livestreamResponse {
	if ls == nil {
		return livestreamResponse{}
	}
	return livestreamResponse{
		ID:                ls.ID,
		WorkspaceID:       ls.WorkspaceID,
		PlatformAccountID: ls.PlatformAccountID,
		Title:             ls.Title,
		Description:       ls.Description,
		PrivacyStatus:     ls.PrivacyStatus,
		PlaybackMode:      ls.PlaybackMode,
		ScheduleType:      ls.ScheduleType,
		ScheduledStartAt:  ls.ScheduledStartAt,
		DesiredState:      ls.DesiredState,
		ActualState:       ls.ActualState,
		Resolution:        ls.Resolution,
		FrameRate:         ls.FrameRate,
		AutoRestart:       ls.AutoRestart,
		Category:          ls.Category,
		MadeForKids:       ls.MadeForKids,
		Language:          ls.Language,
		ThumbnailMediaID:  ls.ThumbnailMediaID,
		DVREnabled:        ls.DVREnabled,
		AutoStart:         ls.AutoStart,
		AutoStop:          ls.AutoStop,
		LatencyPreference: ls.LatencyPreference,
		CreatedAt:         ls.CreatedAt,
		UpdatedAt:         ls.UpdatedAt,
	}
}
